### ISSUE TITLE
Not very different from React

### ISSUE THREAD

**Issue by wmertens** [mark-hahn/space-brush#1]
No way to add comments, so making an issue, sorry.

I'd just like to point out that you can do pretty similar things in pure coffeescript with React, although you can't use the @ syntax. OTOH you get access to the props and state, and the data binding is not magic. I know that React is no longer meant for Atom but still there are elements worth considering...

``` coffeescript
R = React.DOM
N = null

Hello = React.createClass
  getInitialState: -> clicks: 0
  count: ->
    @setState clicks: @state.clicks + 1
  render: ->
    R.div N, "Hello #{@props.name}",
      R.p N, "How are you?"
      R.button onClick: @count, "Click me"
      R.p N, """
        Clicked #{@state.clicks} time#{if @state.clicks != 1 then "s" else ""}.
      """

React.renderComponent Hello(name:"World"), document.body
```

See http://jsfiddle.net/wnocu547/2/ for a live version and http://jlongster.com/Removing-User-Interface-Complexity,-or-Why-React-is-Awesome for a re-implementation of React in 250 lines and http://jackschaedler.github.io/goya/ for a very speedy full-immutable pixel editor with awesome undo/redo thanks to its immutabillity.

Basically I'm saying if Atom can't use React, at least steal its good stuff :grin:.

**Comment by mark-hahn** [comment]
Thanks.  I admit I haven't studied react enough.  I will now.

I had tried to use angular before and hated it.  Especially having so much
crap in the html.  I guess that I'm saying I prefer magic. I am probably
biased against react because I lumped it in with angular.

I did a precursor to space-pen a few years ago in coffeescript.  It was
called drycup but it was replaced by teacup.  Teacup is like space-pen in
that it has no bindings.  So I'm a little (a lot?) biased towards rolling
my own again.

I think my main motivation for this was to try to pressure core into
considering something simple like space-pen again.  When they announced
they were dropping space-pen I went nuts and made a fool out of myself on
atom/atom saying that package developers should havesomething simple. They
announced they were keeping space-pen immediately after that.

So anyway, I will study react  and maybe come back here with some questions.

**Comment by mark-hahn** [comment]
>  OTOH you get access to the props and state

What do you mean by this?  Are you taking about binding to a model?

**Comment by wmertens** [comment]
> > OTOH you get access to the props and state
> 
> What do you mean by this? Are you taking about binding to a model?

In the space-brush examples, the @ makes it easy to denote DOM elements. However in React you can't do that because @ is used to provide the properties of the current DOM element, as well as the state during the lifetime of the element.

I worked on coffeecup and coffeemug a while back and kept thinking it would be cool to re-render only the parts of the application that changed, given a datastructure that you "stretch" across your app. I tried to make that work for a bit but React does it much better :grin:.

Think of React in coffeescript a bit like teacup, only it runs every time you make a change.

**Comment by mark-hahn** [comment]
> I worked on coffeecup and coffeemug a while back and kept thinking it
> would be cool to re-render only the parts of the application that changed,

I did drykup, the module teacup improved on and replaced.  I thought of the
exact same feature but I never actually worked on it.

I'd never hard of coffeemug.  I just checked it out.  He stole my "without
the magic" line.   (grin)

**Comment by wmertens** [comment]
So the main differences between teacup and react are that teacup builds a tree of functions that get called top-down (generally once), and react builds a tree of objects that is constructed bottom-up and compared against the DOM many times. React is also optimized a lot so it only makes the changes it needs to. Check out the rather awesome comparison between Ember, Angular and React here: https://www.youtube.com/watch?v=z5e7kWSHWTg&index=11&list=PLb0IAmt7-GS1cbw4qonlQztYV1TAW0sCr

I have experienced some PHP spaghetti monsters so I was weary of mingling HTML output with code, but with react+(webpack|browserify|requirejs|...) it's in fact totally fine because you can make modular elements that encapsulate only the local domain knowledge, and you can create better code with less jumping around because everything sits right next to each other.

Check out this quick comparison between Angular ("HTML is king") and React: http://codepen.io/wmertens/pen/JovNQo?editors=101 . At first it seems that Angular is nicer with the data binding in HTML without writing any code, but once you need to create elements in a loop and filter arrays etc things get a bit ugly, with controllers having to inject helper functions in the scope and non-javascript mini languages to define loops etc. With React, it starts off a bit uglier but it stays readable and simple no matter how complex your logic is. Helper functions are local to the "HTML" and you can create any "DSL" that best suits your task at hand in a few lines of code.

**Comment by mark-hahn** [comment]
> Check out this quick comparison between Angular ("HTML is king") and
> React: http://codepen.io/wmertens/pen/JovNQo?editors=101 .

I don't need to check out angular.  I know I don't like it.  It's a bloated
monster.

I will check out everything else.  Thanks.

BTW, the Atom core team has announced they are going to pick out one
framework to bless as the one to use by default.

https://github.com/atom/atom/issues/5756

This means almost everyone will use it.  It also means I'm definitely not
going to be using space-brush.  I'm happy.  What motivated me was the lack
of any choice at all.

**Comment by mark-hahn** [comment]
Now that the core team is going to officially support a framework, this whole things is kind of mute.
